### PR TITLE
Add limits to Intl MV

### PR DESCRIPTION
**Pull request by sffc**
Fixes #98

Replaces #117

### CODE REVIEW

File: numberformat/diff.emu [hunk 1]
```diff
@@ -1592,7 +1592,12 @@
         1. Let _text_ be ! StringToCodePoints(_str_).
         1. Let _literal_ be ParseText(_text_, |StringNumericLiteral|).
         1. If _literal_ is a List of errors, return ~not-a-number~.
-        1. Return the StringIntlMV of _literal_.
+        1. Let _intlMV_ be the StringIntlMV of _literal_.
+        1. If _intlMV_ is a mathematical value and _intlMV_ &gt; ℝ(Number.MAX_VALUE), the implementation may set _intlMV_ to ~positive-infinity~.
```
Comment (anba) [inline numberformat/diff.emu:1596]: Nit: % to refer to the original value of `Number.MAX_VALUE`. (Also for the other references below.)

```suggestion
        1. If _intlMV_ is a mathematical value and _intlMV_ &gt; ℝ(%Number.MAX_VALUE%), the implementation may set _intlMV_ to ~positive-infinity~.
```

**Review by anba — APPROVED** [review]

**Review by romulocintra — APPROVED** [review]

**Review by gibson042 — COMMENTED** [review]
I'm not comfortable with the amount of variation this permits within and across implementations. Behavior should really be more predictable.

**Comment by sffc** [comment]
> I'm not comfortable with the amount of variation this permits within and across implementations. Behavior should really be more predictable.

Do you have a counter proposal? Anything I can do to help align on the objective?

**Comment by gibson042** [comment]
Well, can we just remove all variance?
```diff
-        1. If _intlMV_ is a mathematical value and _intlMV_ &gt; ℝ(%Number.MAX_VALUE%), the implementation may set _intlMV_ to ~positive-infinity~.
-        1. If _intlMV_ is a mathematical value and _intlMV_ &lt; ℝ(-%Number.MAX_VALUE%), the implementation may set _intlMV_ to ~negative-infinity~.
-        1. If _intlMV_ is a mathematical value and 0 &lt; _intlMV_ &lt; ℝ(%Number.MIN_VALUE%), the implementation may set _intlMV_ to 0.
-        1. If _intlMV_ is a mathematical value and 0 &gt; _intlMV_ &gt; ℝ(-%Number.MIN_VALUE%), the implementation may set _intlMV_ to ~negative-zero~.
+        1. If _intlMV_ is a mathematical value and _intlMV_ &gt; ℝ(%Number.MAX_VALUE%), set _intlMV_ to ~positive-infinity~.
+        1. If _intlMV_ is a mathematical value and _intlMV_ &lt; ℝ(-%Number.MAX_VALUE%), set _intlMV_ to ~negative-infinity~.
+        1. If _intlMV_ is a mathematical value and 0 &lt; _intlMV_ &lt; ℝ(%Number.MIN_VALUE%), set _intlMV_ to 0.
+        1. If _intlMV_ is a mathematical value and 0 &gt; _intlMV_ &gt; ℝ(-%Number.MIN_VALUE%), set _intlMV_ to ~negative-zero~.
```

If not, then I think the best alternative is introducing symmetric [implementation-defined](https://tc39.es/ecma262/multipage/overview.html#implementation-defined) bounds:
```
        1. If _intlMV_ is a mathematical value and abs(_intlMV_) &gt; MVMagnitudeMaximum(), then
          1. If _intlMV_ &gt; 0, set _intlMV_ to ~positive-infinity~. Otherwise, set _intlMV_ to ~negative-infinity~.
        1. If _intlMV_ is a mathematical value and abs(_intlMV_) &lt; MVMagnitudeMinimum(), then
          1. If _intlMV_ &ge; 0, set _intlMV_ to 0. Otherwise, set _intlMV_ to ~negative-zero~.
```

And in either case, I think it is _also_ necessary to bound the **precision** of these values (e.g., implementations should be allowed—if not required—to treat ``` `1${"0".repeat(100)}.${"0".repeat(100)}1` ``` [`1000…000.000…0001`] identically to ``` `1${"0".repeat(100)}` ``` [`1000…000`]).

**Comment by sffc** [comment]
@gibson042 The rationale for making _soft_ lower bounds is

1. This is already the web reality.
2. It is expensive for browsers to perform these high-order computations in the MV space. For example, V8 fully defers to whatever ICU4C does in this area; the V8 engine does not contain its own BigDecimal-like type to perform this logic.
3. The spec already allows browsers a bit of leeway in the details of how they handle IEEE floating point arithmetic; this is in exactly the same vein.

> And in either case, I think it is also necessary to bound the precision of these values

ECMA-402 requires all numbers to be rounded to a certain number of fraction or significant digits, not more than 20, so that already serves as a de-facto bound on the precision of these MVs.

**Comment by gibson042** [comment]
> 1. This is already the web reality.

I'm saying that users are harmed by the same input being treated as infinite in one implementation but not in another. Why should this proposal not address that?

> 2. It is expensive for browsers to perform these high-order computations in the MV space. For example, V8 fully defers to whatever ICU4C does in this area; the V8 engine does not contain its own BigDecimal-like type to perform this logic.

Implementers are strictly lower in the priority of constituencies than both authors and users. If V8 and/or ICU are failing to meet the needs of the latter groups, then we should push for fixes there (e.g., ICU exposing configuration of limits).

> 3. The spec already allows browsers a bit of leeway in the details of how they handle IEEE floating point arithmetic; this is in exactly the same vein.

I believe you're referring to https://tc39.es/ecma262/multipage/overview.html#implementation-approximated , in which case I dispute the analogy because the difference between e.g. 2<sup>53</sup> and 2<sup>53</sup> + 1 is one of degree while the difference between e.g. 10**308 and Infinity is one of kind.

> ECMA-402 requires all numbers to be rounded to a certain number of fraction or significant digits, not more than 20, so that already serves as a de-facto bound on the precision of these MVs.

Hmm, that's true but it seems brittle.

**Comment by sffc** [comment]
@anba How do you feel about this being a hard limit?

**Comment by sffc** [comment]
> > 1. This is already the web reality.
> 
> I'm saying that users are harmed by the same input being treated as infinite in one implementation but not in another. Why should this proposal not address that?

I argue that the user harm is minimal. The primary motivation behind the Intl MV change was to support large, precise amounts of currency as required by amounts in bitcoin, yen, etc. Supporting more digits than a Number can represent was the pain point. I'm not aware of strong motivating use cases involving large power of 10 exponents.

> > 2. It is expensive for browsers to perform these high-order computations in the MV space. For example, V8 fully defers to whatever ICU4C does in this area; the V8 engine does not contain its own BigDecimal-like type to perform this logic.
> 
> Implementers are strictly lower in the priority of constituencies than both authors and users. If V8 and/or ICU are failing to meet the needs of the latter groups, then we should push for fixes there (e.g., ICU exposing configuration of limits).

While I agree with this in the general sense, when it comes to edge cases, I feel that implementors carry more weight.

This change is motivated by implementation constraints that were discovered in Stage 3. I worry that a hard limit is trading one implementation constraint (internal limits on the size of the number) with another (having to interpret the size of the number).

> > 3. The spec already allows browsers a bit of leeway in the details of how they handle IEEE floating point arithmetic; this is in exactly the same vein.
> 
> I believe you're referring to https://tc39.es/ecma262/multipage/overview.html#implementation-approximated , in which case I dispute the analogy because the difference between e.g. 253 and 253 + 1 is one of degree while the difference between e.g. 10**308 and Infinity is one of kind.

Acknowledged

> > ECMA-402 requires all numbers to be rounded to a certain number of fraction or significant digits, not more than 20, so that already serves as a de-facto bound on the precision of these MVs.
> 
> Hmm, that's true but it seems brittle.

Why is it brittle?

The effective requirement of implementations is that they support ~330 significant digits (~310 before the decimal separator and another ~20 after it for the largest valid value of maximumSignificantDigits). This is another edge case, and it's possible that we don't need so many digits, but any other number of digits is just an arbitrary choice, and it again treads into the territory of implementation challenge.

**Comment by anba** [comment]
> @anba How do you feel about this being a hard limit?

Is there an easy way to compare a string which contains a number in decimal representation against `ℝ(Number.MIN_VALUE)` and `ℝ(Number.MAX_VALUE)`? IOW is there an easy way to get the exact decimal representation of `ℝ(Number.MIN_VALUE)` and `ℝ(Number.MAX_VALUE)`? (When the exact decimal representation is known, we can easily compare the normalised forms against each other. At least as long as we're okay with having to parse the string as exact decimal numbers.)

This is the behaviour before NF v3:
```js
js> new Intl.NumberFormat("en", {minimumSignificantDigits: 1}).format(Number.MIN_VALUE)
"0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005"
js> Number.MIN_VALUE.toPrecision(100)
"4.940656458412465441765687928682213723650598026143247644255856825006755072702087518652998363616359924e-324"
js> new Intl.NumberFormat("en", {minimumSignificantDigits: 1}).format(Number.MIN_VALUE.toPrecision(100))
"0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005"
```

When positive numbers smaller than `ℝ(Number.MIN_VALUE)` are changed to zero, inputs like `"4.940656458412465441765687928682213723650598026143247644255856825006755072702087518652998363616359923e-324"` (`"...4e-324"` was changed to `"...3e-324"`) will have to return `"0"`.

I don't think we care too much about the behaviour of too small or too large values. What if we simply parse the input as a Number and when the result is ±Infinity, ±`Number.MAX_VALUE`, or ±`Number.MIN_VALUE`, take the Number value as the actual input? 

I think this could be expressed through the [RoundMVResult](https://tc39.es/ecma262/#sec-roundmvresult) operation:
> 8. Let intlMV the StringIntlMV of literal.
> 9. If intlMV is a mathematical value, then
>   a. Let rounded be RoundMVResult(abs(intlMV)).
>   b. If rounded is +∞<sub>𝔽</sub> and intlMV < 0, return negative-infinity.
>   c. If rounded is +∞<sub>𝔽</sub>, return positive-infinity.
>   d. If rounded is +0<sub>𝔽</sub> and intlMV < 0, return negative-zero.
>   e. If rounded is +0<sub>𝔽</sub>, return 0.
>   f. If rounded is +%Number.MAX_VALUE% or +%Number.MIN_VALUE%, then
>     &nbsp; &nbsp;     i. If intlMV < 0, return -ℝ(rounded).
>     &nbsp; &nbsp;     ii. Return ℝ(rounded).
> 10. Return intlMV.

**Review by FrankYFTang — APPROVED** [review]
I like the idea of rounding before compare against the limit.

**Review by gibson042 — APPROVED** [review]

**Comment by sffc** [comment]
For record keeping: This PR achieved consensus at the January 2023 TC39-TG1 meeting, the same meeting that the proposal as a whole achieved Stage 4.

https://github.com/tc39/notes/blob/main/meetings/2023-01/jan-31.md#intlnumberformat-v3-for-stage-4

I said in my presentation:

> In interpret strings as decimals, this is one thing I’m highlighting a change that happened since November 2022 at the end of the last meeting. I asked for some feedback on ways to go about this problem of how the set the range of the allowed Intl math mat value. I took that feedback and merged it with feedback from implementers, and this is the statement that we ended up with. So Intl mathematical values have range equal to a number, meaning that the magnitude of numbers represented are as equal to the range supported by a number type. But greater precision is allowed. So you can have more significant digits. The maximum number of significant digits is enforced by the fact that Intl number format always rounds numbers to a certain -- to a limited precision or number of fractional digits.

The corresponding slide is:

https://docs.google.com/presentation/d/1b627TYDVDDcdae9D80CP5DLnSYX8GU97nmritxVN5Wo/edit?slide=id.g82ae0c50ed_0_111#slide=id.g82ae0c50ed_0_111

I had previously raised the question at the November 2022 TC39 meeting:

https://github.com/tc39/notes/blob/main/meetings/2022-11/nov-29.md#intl-numberformat-v3

**Pull request merged**